OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_VULKAN_LAYER_COMMAND_BUFFER_MANAGER_H_
+#define ORBIT_VULKAN_LAYER_COMMAND_BUFFER_MANAGER_H_
+
+#include <stack>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Profiling.h"
+#include "VulkanLayerProducer.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "vulkan/vulkan.h"
+
+namespace orbit_vulkan_layer {
+
+namespace internal {
+
+// Stores meta information about a submit (VkQueueSubmit), e.g. to allow to identify the matching
+// Gpu tracepoint.
+struct SubmissionMetaInformation {
+  uint64_t pre_submission_cpu_timestamp;
+  uint64_t post_submission_cpu_timestamp;
+  int32_t thread_id;
+};
+
+// A persistent version of a command buffer that was submitted and its begin/end slot in the
+// `TimerQueryPoool`. Note that the begin is optional, as it might not be part of the capture.
+// This struct is created if we capture the submission, so if we would not have captured the end
+// we also not have captured the begin, and don't need to store info about that command buffer at
+// all.
+struct SubmittedCommandBuffer {
+  std::optional<uint32_t> command_buffer_begin_slot_index;
+  uint32_t command_buffer_end_slot_index;
+};
+
+// A persistent version of a debug marker (either begin or end) that was submitted and its slot
+// in the `TimerQueryPool`. Note that we also store the `SubmissionMetaInformation` as begin markers
+// can originate in different submissions then the matching end markers.
+struct SubmittedMarker {
+  SubmissionMetaInformation meta_information;
+  uint32_t slot_index;
+};
+
+// Represents a color to be used to in debug markers. The values are all in range [0.f, 1.f.].
+struct Color {
+  float red;
+  float green;
+  float blue;
+  float alpha;
+};
+
+// Identifies a particular debug marker region. We have a stack of all markers of a queue, that gets
+// updated upon a submission (VkQueueSubmit). If at that time we have a value for the end_info, we
+// use that structure to persist the marker description. So once it is submitted, the
+// end_info will always set.
+// Beside the information about the begin/end, it also stores the text, color and the depth of the
+// marker. If a debug marker begin was limited because of its depth, cut_off is set to true.
+// This allows end markers on a different submission to also through the end marker away.
+// Example: Max Depth = 1
+// Submission 1: Begin("Foo"), Begin("Bar) -- "Bar" will be cut-off
+// Submission 2: End("Bar"), End("Foo") -- We now know, that the first end needs to be cut-off.
+struct MarkerState {
```

Comment:
`MarkerState` is used in two contexts.
1. There is a stack `QueueMarkerState` for each queue as internal state of SubmissionTracker, that keeps track of the markers on each submission. That was the reason, why I wanted to name it "State" here. Because, it will change.
On a submitted "Begin" marker, we will create that state, fill the information for the begin and push it to that stack. At the "End", it will be popped, and the information of the "end" will be added. If we are capturing, that information will be stored in the `QueueSubmission` as `completed_markers`. 

2. Is this use-case. Here it is not modified and acts more as a "Slice".

Reply:
I extended the comment here.

Reply:
Would it be better or worse to have two different `struct`s for the two uses? I think you would also have slightly different fields, wouldn't you?

Reply:
I also thought of this. The only difference would be that the "end_info" wouldn't be an optional. So I wanted to save the extra copy and allow them to move. WDYT?

Reply:
If the reason is performance, I think that the only member that is actually moved is the string (which you can still move explicitly when converting from a struct to another), everything else is copied as they are primitive types or structs of primitive types.

Reply:
But I'm still confused about `cut_off`. Is that carried over?